pkg/algorithm/cell_allocation.go

```diff
@@ -27,119 +27,200 @@ import (
 	"github.com/microsoft/hivedscheduler/pkg/api"
 	"github.com/microsoft/hivedscheduler/pkg/common"
 	"k8s.io/klog"
-	"math"
-	"math/rand"
+	"sort"
 )
 
 // buddyAlloc is used for allocating a free physical cell to a preassigned virtual cell.
 // It splits a higher-level cell when there is no free cell at the current level.
-// As the input cell list is a copy of the real free list and hence is one-off,
-// we won't remove a returned cell from it.
-func buddyAlloc(freeList ChainCellList, level CellLevel, suggestedNodes common.Set) *PhysicalCell {
-	if len(freeList[level]) == 0 && level < CellLevel(len(freeList)) {
-		higherCell := buddyAlloc(freeList, level+1, suggestedNodes)
-		if higherCell != nil {
-			freeList[level] = append(freeList[level], higherCell.GetChildren()...)
+// Note that this is a backtracking version of the buddy alloc algorithm (slightly more complex
+// than the simple recursive version present in the paper). Ideally we do not need to backtrack
+// because buddy alloc has safety guarantee (the simple algorithm already works).
+// But it's possible that the cell allocated by buddy alloc is temporarily unavailable
```

Comment:
>cell allocated by buddy alloc is temporarily unavailable [](start = 30, length = 56)

cell allocated by non-backtracking version buddy alloc is temporarily unavailable